sail/src/main/java/org/apache/rya/rdftriplestore/inference/InferenceEngine.java

```diff
@@ -416,22 +425,131 @@ private void refreshHasValueRestrictions(Map<Resource, URI> restrictions) throws
+                    ryaDaoQueryWrapper.queryFirst(head, RDF.FIRST, null, new RyaDaoStatementIterHandler() {
+                        @Override
+                        public void handleStatementIter(final Statement st2) throws Exception{
+                            // The object found in the query represents a type
+                            // that should be included in the intersection.
+                            final URI obj2 = (URI) st2.getObject();
+                            intersection.add(obj2);
+                        }
+                    });
+                    final List<URI> headHolder = new ArrayList<>(1);
+                    // rdf.rest will point to the next bnode that's part of the
+                    // intersection.
+                    ryaDaoQueryWrapper.queryFirst(head, RDF.REST, null, new RyaDaoStatementIterHandler() {
+                        @Override
+                        public void handleStatementIter(final Statement st3) throws Exception {
+                            // This object is the next bnode head to look for.
+                            final URI obj3 = (URI) st3.getObject();
+                            headHolder.add(obj3);
+                        }
+                    });
+                    // As long as we get a new head there are more bnodes that
+                    // are part of the intersection. Keep going until we reach
+                    // rdf.nil.
+                    if (!headHolder.isEmpty()) {
+                        head = headHolder.get(0);
+                    } else {
+                        head = RDF.NIL;
+                    }
+                }
+                // Add this intersection for this type. There may be more
+                // intersections for this type so each type has a list of
+                // intersection sets.
+                intersectionsProp.get(type).add(intersection);
+            }
+        });
+
+        for (final Map.Entry<Resource, List<Set<Resource>>> entry : intersectionsProp.entrySet()) {
+            final Resource type = entry.getKey();
+            final List<Set<Resource>> intersectionList = entry.getValue();
+            final Set<Resource> otherTypes = new HashSet<>();
+            // Combine all of a type's intersections together.
+            for (final Set<Resource> intersection : intersectionList) {
+                otherTypes.addAll(intersection);
+            }
+            for (final Resource other : otherTypes) {
+                // :A intersectionOf[:B, :C] implies that
+                // :A subclassOf :B
+                // :A subclassOf :C
+                // So add each type that's part of the intersection to the
+                // subClassOf graph.
+                addSubClassOf(type, other);
+                for (final Set<Resource> intersection : intersectionList) {
+                    if (!intersection.contains(other)) {
+                        addIntersection(intersection, other);
```

Comment:
I'm not sure if that's the same. I can read "A = (B and C)" two ways -- either "A is the intersection of B and C":
    <A> owl:intersectionOf (<B> <C>) .
    <D> owl:intersectionOf (<B> <C>) .
    <D> rdfs:subClassOf <E> .
or "A is equivalent to some anonymous class which is the intersection of B and C:
    <A> owl:equivalentClass _:bnode
    _:bnode owl:intersectionOf (<B> <C>) .
    _:bnode owl:equivalentClass <D> .
    <D> rdfs:subClassOf <E> .

In the latter case, we have A subClassOf E just from subclass and equivalent class transitivity, without touching the intersection.

In the former, I suppose you could infer that A and D are equivalent classes by comparing their intersections. That would require comparing the actual contents of all intersection lists to find any pairs that contain the same set of classes. That might be something to do in the future, but I don't think I'd try to do it at this point since we're not making that kind of check with any of the other class expressions (unions, property restrictions, etc).